src/openrct2/world/MapGen.cpp

```diff
@@ -259,52 +258,64 @@ static void mapgen_place_tree(sint32 type, sint32 x, sint32 y)
  */
 static void mapgen_place_trees()
 {
-    sint32 numGrassTreeIds = 0, numDesertTreeIds = 0, numSnowTreeIds = 0;
-    sint32 *grassTreeIds = (sint32*)malloc(countof(GrassTrees) * sizeof(sint32));
-    sint32 *desertTreeIds = (sint32*)malloc(countof(DesertTrees) * sizeof(sint32));
-    sint32 *snowTreeIds = (sint32*)malloc(countof(SnowTrees) * sizeof(sint32));
+    std::vector<sint32> grassTreeIds(Util::CountOf(GrassTrees), 0);
+    std::vector<sint32> desertTreeIds(Util::CountOf(DesertTrees), 0);
+    std::vector<sint32> snowTreeIds(Util::CountOf(SnowTrees), 0);
 
-    for (sint32 i = 0; i < object_entry_group_counts[OBJECT_TYPE_SMALL_SCENERY]; i++) {
-        rct_scenery_entry *sceneryEntry = get_small_scenery_entry(i);
-        rct_object_entry_extended *entry = &object_entry_groups[OBJECT_TYPE_SMALL_SCENERY].entries[i];
+    for (sint32 i = 0; i < object_entry_group_counts[OBJECT_TYPE_SMALL_SCENERY]; i++)
+    {
+        rct_scenery_entry         * sceneryEntry = get_small_scenery_entry(i);
+        rct_object_entry_extended * entry        = &object_entry_groups[OBJECT_TYPE_SMALL_SCENERY].entries[i];
 
-        if (sceneryEntry == NULL)
+        if (sceneryEntry == nullptr)
             continue;
 
-        sint32 j;
-        for (j = 0; j < countof(GrassTrees); j++)
+        uint32 j;
+        for (j = 0; j < Util::CountOf(GrassTrees); j++)
+        {
             if (strncmp(GrassTrees[j], entry->name, 8) == 0)
                 break;
-        if (j != countof(GrassTrees)) {
-            grassTreeIds[numGrassTreeIds++] = i;
+        }
+        if (j != Util::CountOf(GrassTrees))
+        {
+            grassTreeIds.push_back(i);
             continue;
         }
 
-        for (j = 0; j < countof(DesertTrees); j++)
+        for (j = 0; j < Util::CountOf(DesertTrees); j++)
+        {
             if (strncmp(DesertTrees[j], entry->name, 8) == 0)
                 break;
-        if (j != countof(DesertTrees)) {
-            desertTreeIds[numDesertTreeIds++] = i;
+        }
+        if (j != Util::CountOf(DesertTrees))
+        {
+            desertTreeIds.push_back(i);
             continue;
         }
 
-        for (j = 0; j < countof(SnowTrees); j++)
+        for (j = 0; j < Util::CountOf(SnowTrees); j++)
+        {
             if (strncmp(SnowTrees[j], entry->name, 8) == 0)
                 break;
-        if (j != countof(SnowTrees)) {
-            snowTreeIds[numSnowTreeIds++] = i;
+        }
+        if (j != Util::CountOf(SnowTrees))
+        {
+            snowTreeIds.push_back(i);
             continue;
         }
     }
 
     sint32 availablePositionsCount = 0;
-    struct { sint32 x; sint32 y; } tmp, *pos, *availablePositions;
-    availablePositions = malloc(MAXIMUM_MAP_SIZE_TECHNICAL * MAXIMUM_MAP_SIZE_TECHNICAL * sizeof(tmp));
+    LocationXY32 tmp, * pos;
+
+    std::vector<LocationXY32> availablePositions(MAXIMUM_MAP_SIZE_TECHNICAL * MAXIMUM_MAP_SIZE_TECHNICAL * sizeof(tmp));
```

Comment:
Remove capacity, so just `std::vector<LocationXY32> availablePositions`
Remove availablePositionsCount, instead use `availablePositions.size()`
Make `pos` not a pointer.
Add new positions to the vector using `push_back`